protostar/cheatable_starknet/controllers/expect_events_controller.py

```diff
@@ -0,0 +1,166 @@
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Optional
+
+from protostar.starknet import Address, CairoData, Selector, ReportedException
+
+if TYPE_CHECKING:
+    from protostar.testing import Hook
+    from protostar.cairo_testing import CairoTestExecutionState
+    from protostar.cheatable_starknet.cheatables.cheatable_cached_state import (
+        CheatableCachedState,
+    )
+
+
+@dataclass
+class Event:
+    from_address: Address
+    key: Selector
+    data: Optional[CairoData] = None
+
+
+@dataclass
+class EventMatching:
+    pass
+
+
+@dataclass
+class SkippedEventMatching(EventMatching):
+    emitted_event: Event
+
+
+@dataclass
+class AcceptedEventMatching(EventMatching):
+    emitted_event: Event
+    expected_event: Event
+
+
+@dataclass
+class FailedEventMatching(EventMatching):
+    expected_event: Event
+
+
+@dataclass
+class EventMatchingResult:
+    event_matchings: list[EventMatching]
+
+    @property
+    def failed_event_matchings(self):
+        return [
+            matching
+            for matching in self.event_matchings
+            if isinstance(matching, FailedEventMatching)
+        ]
+
+    @property
+    def should_be_accepted(self) -> bool:
+        return not self.failed_event_matchings
```

Comment:
```suggestion
@dataclass
class EventMatchingResult:
    accepted_event_matchings: list[AcceptedEventMatching]
    failed_event_matchings: list[FailedEventMatching]
    skipped_event_matchings: list[SkippedEventMatching]

    @property
    def should_be_accepted(self) -> bool:
        return not self.failed_event_matchings
```

How about storing specific results in separate collections rather than querying them all the time like this and spawning new comprehension lists?

Reply:
We need to maintain the order to display matchings correctly.